include/substrait/type/Type.h

```diff
@@ -0,0 +1,701 @@
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  const ParameterizedTypePtr elementType_;
+};
+
+class ParameterizedStruct : public ParameterizedTypeBase {
+ public:
+  explicit ParameterizedStruct(
+      std::vector<ParameterizedTypePtr> types,
+      bool nullable = false)
+      : ParameterizedTypeBase(nullable), children_(std::move(types)) {}
+
+  [[nodiscard]] std::string signature() const override;
+
+  [[nodiscard]] const std::vector<ParameterizedTypePtr>& children() const {
+    return children_;
+  }
+
+  [[nodiscard]] TypeKind kind() const override {
+    return TypeKind::kStruct;
+  }
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  const std::vector<ParameterizedTypePtr> children_;
+};
+
+class ParameterizedMap : public ParameterizedTypeBase {
+ public:
+  ParameterizedMap(
+      ParameterizedTypePtr keyType,
+      ParameterizedTypePtr valueType,
+      bool nullable = false)
+      : ParameterizedTypeBase(nullable),
+        keyType_(std::move(keyType)),
+        valueType_(std::move(valueType)) {}
+
+  [[nodiscard]] const ParameterizedTypePtr& keyType() const {
+    return keyType_;
+  }
+
+  [[nodiscard]] TypeKind kind() const override {
+    return TypeKind::kMap;
+  }
+  [[nodiscard]] const ParameterizedTypePtr& valueType() const {
+    return valueType_;
+  }
+
+  [[nodiscard]] std::string signature() const override;
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  const ParameterizedTypePtr keyType_;
+  const ParameterizedTypePtr valueType_;
+};
+
+std::shared_ptr<const ScalarType<TypeKind::kBool>> BOOL();
+
+std::shared_ptr<const ScalarType<TypeKind::kI8>> TINYINT();
+
+std::shared_ptr<const ScalarType<TypeKind::kI16>> SMALLINT();
+
+std::shared_ptr<const ScalarType<TypeKind::kI32>> INTEGER();
+
+std::shared_ptr<const ScalarType<TypeKind::kI64>> BIGINT();
+
+std::shared_ptr<const ScalarType<TypeKind::kFp32>> FLOAT();
+
+std::shared_ptr<const ScalarType<TypeKind::kFp64>> DOUBLE();
+
+std::shared_ptr<const ScalarType<TypeKind::kString>> STRING();
+
+std::shared_ptr<const ScalarType<TypeKind::kBinary>> BINARY();
+
+std::shared_ptr<const ScalarType<TypeKind::kTimestamp>> TIMESTAMP();
+
+std::shared_ptr<const ScalarType<TypeKind::kTimestampTz>> TIMESTAMP_TZ();
+
+std::shared_ptr<const ScalarType<TypeKind::kDate>> DATE();
+
+std::shared_ptr<const ScalarType<TypeKind::kTime>> TIME();
+
+std::shared_ptr<const ScalarType<TypeKind::kIntervalYear>> INTERVAL_YEAR();
+
+std::shared_ptr<const ScalarType<TypeKind::kIntervalDay>> INTERVAL_DAY();
+
+std::shared_ptr<const ScalarType<TypeKind::kUuid>> UUID();
+
+std::shared_ptr<const Decimal> DECIMAL(int precision, int scale);
+
+std::shared_ptr<const Varchar> VARCHAR(int len);
+
+std::shared_ptr<const FixedChar> FChar(int len);
+
+std::shared_ptr<const FixedBinary> FBinary(int len);
```

Comment:
For consistency should this be `FCHAR`/`FBINARY` or `FIXED_CHAR`/`FIXED_BINARY`?

Reply:
updated